### PR TITLE
Add PFNs4BO sampler

### PR DESCRIPTION
**Pull request by HideakiImamura**
## Contributor Agreements

Please read the [contributor agreements](https://github.com/optuna/optunahub-registry/blob/main/CONTRIBUTING.md#contributor-agreements) and if you agree, please click the checkbox below.

- [x] I agree to the contributor agreements.

## Motivation

PFNs4BO was recently proposed by Samuel Müller, Matthias Feurer, Noah Hollmann, and Frank Hutter in https://arxiv.org/abs/2305.17535. It is highly interesting algorithm and can be applied to any BO problem. This PR aims to add the method in the OptunaHub.

## Description of the changes

- Add PFNs4BO sampler

### CODE REVIEW

**Comment by HideakiImamura** [comment]
@y0z Could you review this PR?

**Comment by HideakiImamura** [comment]
I conducted benchmark experiments of PFNs4BO against the random search. PFNs4BO looks promising.

compare2-pfns4bo-vs-random-1

This figure stands for the performance of PFNs4BO against the random search. The horizontal axis represents the amount of  budgets, which is equal to the number of trials. The vertical axis represents the p-value of the Mann–Whitney U test, in which the null hypothesis is that the PFNs4BO loses the random search. Therefore, don't get me wrong, but the lower is better.

The benchmark consists of 70 problems. The 8 problems are from HPO tabular benchmarks in https://github.com/automl/HPOBench (original 4 problems, and log-converted 4 problems). The 62 problems are several synthetic functions in https://github.com/sigopt/evalset (Not just use it, but I re-implement them).

File: package/samplers/pfns4bo/sampler.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+        self._independent_sampler = independent_sampler or RandomSampler(seed=seed)
+        self._intersection_search_space = IntersectionSearchSpace()
+        self._n_startup_trials = n_startup_trials
+
+        self._device = utils.default_device
+
+        if isinstance(prior, torch.nn.Module):
+            trained_model = prior
+        elif prior == "vanilla gp":
+            _, _, trained_model, _ = train(**get_vanilla_gp_config(self._device))
```
Comment (SamuelGabriel) [inline package/samplers/pfns4bo/sampler.py:169]: I am not sure what your application is exactly. But training a PFN like the pre-trained ones takes around 1 day *with a GPU*. Thus, I don't think it is realistic HPO users want to do this, but they probably just want to use a pre-trained model. Either the ones I provided, or if you like to train better ones, then these.

File: package/samplers/pfns4bo/sampler.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+        states = (TrialState.COMPLETE,)
+        trials = study._get_trials(deepcopy=False, states=states, use_cache=True)
+
+        if len(trials) < self._n_startup_trials:
+            return {}
+
+        (
+            internal_search_space,
+            normalized_params,
+        ) = gp_search_space.get_search_space_and_normalized_params(trials, search_space)
```
Comment (SamuelGabriel) [inline package/samplers/pfns4bo/sampler.py:198]: The normalization of the HPs should be to [0,1] interval. Not sure, if that is what this function does. Just as a note :)

File: package/samplers/pfns4bo/sampler.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+            _, x_options, eis, _, _ = optimize_acq_w_lbfgs(
+                model=self._model,
+                known_x=known_x,
+                known_y=known_y,
+                num_grad_steps=self._num_grad_steps,
+                num_candidates=self._num_candidates,
+                pre_sample_size=self._pre_sample_size,
+                device=self._device,
+                rand_sample_func=rand_sample_func,
+                apply_power_transform=False,
```
Comment (SamuelGabriel) [inline package/samplers/pfns4bo/sampler.py:235]: for us, power_transform was important to get really good scores. Maybe you can make it a hyper-parameter at least.

File: package/samplers/pfns4bo/sampler.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+        def rand_sample_func(n: int) -> torch.Tensor:
+            xs = sample_normalized_params(n, internal_search_space, None)
+            ret = torch.from_numpy(xs).to(torch.float32).to(self._device)
+            return ret
+
+        known_x = torch.from_numpy(normalized_params).to(torch.float32).to(self._device)
+        known_y = torch.from_numpy(standarized_score_vals).to(torch.float32).to(self._device)
+
+        with torch.enable_grad():
+            _, x_options, eis, _, _ = optimize_acq_w_lbfgs(
```
Comment (SamuelGabriel) [inline package/samplers/pfns4bo/sampler.py:226]: You can also pass an arg `acq_function` here, to specify to use another acquisition function from the ones specified in https://github.com/automl/PFNs4BO/blob/486e1ad36cb2d8e1ad933486a829e64395161b85/pfns4bo/scripts/acquisition_functions.py#L267 (some of these might not have gradients, though, but at least PI should also work)

**Review by SamuelGabriel — COMMENTED** [review]
To me this looks generally good. I think using power transform would improve performance further and you could also have something in place to use random search after 100 or 200 trials, as PFNs4BO becomes worse then.

Other than that: I did not double check all hyperparameters, but if you use the ones in the code (which are the same as in the paper), you should be fine.

File: package/samplers/pfns4bo/sampler.py (hunk 2 above)
Comment (HideakiImamura) [reply package/samplers/pfns4bo/sampler.py:198]: Thanks for the comment. Yes, this search space transformation allow us to transform each variable into [0,1] interval.

File: package/samplers/pfns4bo/sampler.py (hunk 1 above)
Comment (HideakiImamura) [reply package/samplers/pfns4bo/sampler.py:169]: Sounds  very good. I completely agree with you. Since we should provide the pre-trained model for users, I will change the default value of the prior to use that provided in [here](https://github.com/automl/PFNs/blob/main/models_diff/prior_diff_real_checkpoint_n_0_epoch_42.cpkt).

File: package/samplers/pfns4bo/sampler.py (hunk 3 above)
Comment (HideakiImamura) [reply package/samplers/pfns4bo/sampler.py:235]: Thank you for your input. I change the default value of it True.

File: package/samplers/pfns4bo/sampler.py (hunk 4 above)
Comment (HideakiImamura) [reply package/samplers/pfns4bo/sampler.py:226]: Thank you for the comment. It would be great for users to specify the type of acquisition functions. I will add the argument of `acquisition_function_type` to specify the type.

**Comment by HideakiImamura** [comment]
@SamuelGabriel Thank you for providing several important comments. I addressed all of your suggestions. Thank you vary much.

> using power transform would improve performance further

I will use `apply_power_transform = True` in the acquisition function optimization. Thanks for the information.

> you could also have something in place to use random search after 100 or 200 trials, as PFNs4BO becomes worse then.

Sounds very good. That type of algorithm change is easily possible in Optuna. Users can change the algorithm as follows. I will add that instruction in the docstring and README.

```python
sampler1 = PFNsBOSampler()
study = optuna.create_study(sampler=sampler)

study.optimize(objective, n_trials=200)

sampler2 = RandomSampler()
study.sampler = sampler2
study.optimize(objective, n_trials=100)
```


> Other than that: I did not double check all hyperparameters, but if you use the ones in the code (which are the same as in the paper), you should be fine.

I used completely same hyperparameters! Thank you for the information.

**Comment by HideakiImamura** [comment]
@y0z Could you review this PR?

File: package/samplers/pfns4bo/README.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
+---
+author: "HideakiImamura"
```
Comment (y0z) [inline package/samplers/pfns4bo/README.md:2]: Don't need white space? (i.e., Hideaki Imamura)

File: package/samplers/pfns4bo/README.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+title: "PFNs4BO sampler"
+description: "In-context learning for Bayesian optimization. This sampler uses Prior-data Fitted Networks (PFNs) as a surrogate model for Bayesian optimization."
+tags: ["sampler"]
```
Comment (y0z) [inline package/samplers/pfns4bo/README.md:5]: Why not add tags to help users find your sampler?

File: package/samplers/pfns4bo/README.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+optuna_versions: ["3.6.1"]
+license: "MIT License"
+---
+
+## Class or Function Names
+- PFNs4BOSampler
+
+## Installation
+```bash
+pip install -r requirements.txt
```
Comment (y0z) [inline package/samplers/pfns4bo/README.md:15]: The following will be more helpful because this allows users to install the dependencies by copy and paste. The URL becomes valid after merging this PR.

```suggestion
pip install -r https://raw.githubusercontent.com/optuna/optunahub-registry/main/package/samplers/pfns4bo/requirements.txt
```

File: package/samplers/pfns4bo/README.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+The default prior argument is ``"hebo"``. This trains the PFNs model in the init of the sampler. If you want to use a pre-trained model, you can download the model checkpoint from the following link: https://github.com/automl/PFNs/blob/main/models_diff/prior_diff_real_checkpoint_n_0_epoch_42.cpkt and load it using the following code:
+
+```python
+import torch
+
+model = torch.load("PATH/TO/prior_diff_real_checkpoint_n_0_epoch_42.cpkt")
+sampler = PFNs4BOSampler(prior=model)
+```
+
+The performance of PFNs4BO with the HEBO+ prior is maximized with the number of trials smaller than 100 or 200 in most cases. If you have a large number of trials, it is recommended to change the sampler to a random sampler or etc after a certain  number of trials.
```
Comment (y0z) [inline package/samplers/pfns4bo/README.md:63]:
```suggestion
The performance of PFNs4BO with the HEBO+ prior is maximized with the number of trials smaller than 100 or 200 in most cases. If you have a large number of trials, changing the sampler to another one (e.g., a random sampler) after a certain number of trials is recommended.
```

File: package/samplers/pfns4bo/README.md [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+### Reference
+
+Samuel Müller, Matthias Feurer, Noah Hollmann, and Frank Hutter. 2023. PFNs4BO: in-context learning for Bayesian optimization. In Proceedings of the 40th International Conference on Machine Learning (ICML'23), Vol. 202. JMLR.org, Article 1056, 25444–25470.
```
Comment (y0z) [inline package/samplers/pfns4bo/README.md:67]: It seems helpful to add a link to [the paper](https://arxiv.org/abs/2305.17535).

**Review by y0z — COMMENTED** [review]
I leave some review comments.

File: package/samplers/pfns4bo/README.md (hunk 7 above)
Comment (HideakiImamura) [reply package/samplers/pfns4bo/README.md:15]: I think resolving dependencies might fail when the environment of users do not match the latest main branch. It would be great as it is.

**Review by y0z — APPROVED** [review]
LGTM

File: package/samplers/pfns4bo/sampler.py (hunk 1 above)
Comment (SamuelGabriel) [reply package/samplers/pfns4bo/sampler.py:227]: No these: https://github.com/automl/PFNs4BO/tree/main/pfns4bo/final_models
